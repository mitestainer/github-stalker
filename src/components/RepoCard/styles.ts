import styled, { css } from 'styled-components'

export const Wrapper = styled.a`
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
    border-radius: 10px;
    overflow: hidden;
    cursor: pointer;
    display: flex;
    flex-direction: column;
    row-gap: 1rem;
    padding: ${theme.gutter};
    border: 1px solid ${theme.colors.lightgray};
    height: 17rem;
    text-decoration: none;
  `}
`

export const Content = styled.div`
  text-align: left;
  flex: 1 1;
`

export const Title = styled.p`
  ${({ theme }) => css`
    font-weight: ${theme.font.weight.bold};
    color: ${theme.colors.gray};
    margin-bottom: 10px;
  `}
`

export const Description = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.size.small};
    color: ${theme.colors.black};
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: normal;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 3;
  `}
`

export const Footer = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: flex-end;
    font-size: ${theme.font.size.xsmall};
    column-gap: 1rem;
  `}
`

export const UpdatedAt = styled.span`
  ${({ theme }) => css`
    font-size: ${theme.font.size.xxsmall};
    color: ${theme.colors.gray};
    flex: 1 0;
  `}
`

export const Star = styled.span`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    color: ${theme.colors.gray};

    > svg {
      fill: ${theme.colors.lightgray};
      color: ${theme.colors.gray};
      margin-right: 0.3rem;
    }
  `}
`

export const Badge = styled.span`
  ${({ theme }) => css`
    background-color: ${theme.colors.purple};
    padding: 5px 10px;
    border-radius: ${theme.borderRadius};
    border: 1px solid ${theme.colors.darkPurple};
    color: ${theme.colors.darkPurple};
  `}
`
