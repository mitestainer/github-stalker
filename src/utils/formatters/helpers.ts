export const formatDate = (dateString: string) => {
  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(new Date(dateString))
}
