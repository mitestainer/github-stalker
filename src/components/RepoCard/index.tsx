import { FiStar } from 'react-icons/fi'
import * as S from './styles'

export type RepoCardProps = {
  name: string
  description: string
  stars: number
  language?: string
  url: string
  updatedAt: string
}

const RepoCard = ({
  name,
  description,
  stars,
  language,
  url,
  updatedAt
}: RepoCardProps) => {
  return (
    <S.Wrapper href={url} target="_blank">
      <S.Content>
        <S.Title>{name}</S.Title>
        <S.Description>{description}</S.Description>
      </S.Content>
      <S.Footer>
        <S.UpdatedAt>Last update: {updatedAt}</S.UpdatedAt>
        <S.Star aria-label="stars">
          <FiStar />
          {stars}
        </S.Star>
        {!!language && <S.Badge aria-label="language">{language}</S.Badge>}
      </S.Footer>
    </S.Wrapper>
  )
}

export default RepoCard
