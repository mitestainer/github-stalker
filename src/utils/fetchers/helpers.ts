import { formatDate } from 'utils/formatters/helpers'

type RepoResponseProps = {
  id: number
  full_name: string
  description: string
  html_url: string
  watchers: number
  language: string
  updated_at: string
}

export const fetchRepos = async (handle: string, page: number) => {
  const response = await fetch(
    `https://api.github.com/orgs/${handle}/repos?sort=updated&direction=desc&page=${page}&per_page=12`
  )
  const data = await response.json()

  return data.map((repo: RepoResponseProps) => ({
    id: repo.id,
    name: repo.full_name,
    description: repo.description,
    url: repo.html_url,
    stars: repo.watchers,
    language: repo.language,
    updatedAt: formatDate(repo.updated_at)
  }))
}
