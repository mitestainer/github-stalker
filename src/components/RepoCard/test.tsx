import { render, screen } from 'utils/tests/helpers'

import RepoCard from '.'

const mockData = {
  name: 'React',
  description: 'The repository for React',
  stars: 1234,
  language: 'JavaScript',
  url: 'https://github.com/facebook/react',
  updatedAt: '2021-11-14T14:03:02Z'
}

describe('<RepoCard />', () => {
  it('should render the RepoCard component and its content', () => {
    const { container } = render(<RepoCard {...mockData} />)

    expect(container.firstChild).toHaveAttribute('href', mockData.url)
    expect(screen.getByText(mockData.name)).toBeInTheDocument()
    expect(screen.getByText(mockData.name)).toHaveStyleRule('color', '#8383a6')
    expect(screen.getByText(mockData.description)).toBeInTheDocument()
    expect(screen.getByLabelText(/stars/i)).toBeInTheDocument()
    expect(screen.getByText(mockData.stars)).toBeInTheDocument()
    expect(screen.getByLabelText(/language/i)).toBeInTheDocument()
    expect(screen.getByText(mockData.language)).toBeInTheDocument()
    expect(screen.getByText(mockData.language)).toHaveStyleRule(
      'background-color',
      '#9387e1'
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should not render the language label if the repo has no language', () => {
    render(<RepoCard {...mockData} language="" />)

    expect(screen.queryByLabelText(/language/i)).not.toBeInTheDocument()
  })
})
