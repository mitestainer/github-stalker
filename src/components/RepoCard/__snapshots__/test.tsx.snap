// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<RepoCard /> should render the RepoCard component and its content 1`
] = `
.c0 {
  background-color: #fff;
  border-radius: 10px;
  overflow: hidden;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 1rem;
  padding: 2rem;
  border: 1px solid #dadae5;
  height: 17rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 {
  text-align: left;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
}

.c2 {
  font-weight: 600;
  color: #8383a6;
  margin-bottom: 10px;
}

.c3 {
  font-size: 1.4rem;
  color: #000;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: normal;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  font-size: 1.2rem;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
}

.c5 {
  font-size: 1rem;
  color: #8383a6;
  -webkit-flex: 1 0;
  -ms-flex: 1 0;
  flex: 1 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #8383a6;
}

.c6 > svg {
  fill: #dadae5;
  color: #8383a6;
  margin-right: 0.3rem;
}

.c7 {
  background-color: #9387e1;
  padding: 5px 10px;
  border-radius: 1rem;
  border: 1px solid #5f49ca;
  color: #5f49ca;
}

<a
  class="c0"
  href="https://github.com/facebook/react"
  target="_blank"
>
  <div
    class="c1"
  >
    <p
      class="c2"
    >
      React
    </p>
    <p
      class="c3"
    >
      The repository for React
    </p>
  </div>
  <div
    class="c4"
  >
    <span
      class="c5"
    >
      Last update: 
      2021-11-14T14:03:02Z
    </span>
    <span
      aria-label="stars"
      class="c6"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
      1234
    </span>
    <span
      aria-label="language"
      class="c7"
    >
      JavaScript
    </span>
  </div>
</a>
`
